documentation/writing_tests.md

```diff
@@ -0,0 +1,949 @@
+- `after` and `afterEach` should be used to tear down things that have been created in `before` and `beforeEach`.
+
+Code in either of these methods can be abstracted away into utility functions that may live outside of each test file **only if** these pieces of code are relevant to more than one test file.
+
+One specific example of code that **must** live in the `beforeEach` section is [the recorder](#the-recorder)'s initialization, and later in the `afterEach` section, the recorder's destruction. The following example can be seen in [the recorder's readme](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/test-utils/recorder/README.md):
+
+```ts
+  beforeEach(async function() {
+    recorder = record(that);
+    const credential = await new ClientSecretCredential(
+      env.AZURE_TENANT_ID,
+      env.AZURE_CLIENT_ID,
+      env.AZURE_CLIENT_SECRET
+    );
+
+    const keyVaultUrl = "https://myKeyVault.vault.azure.net";
+    client = new KeysClient(keyVaultUrl, credential);
+  });
+
+  afterEach(function () {
+    recorder.stop();
+  });
+```
+
+### Preparing some of the test cases
+
+Some test cases might need some common setup to properly work. If at least more than one of these tests exist in a single file, their setup can be moved out of each test case into a function (or outside constant), as long as the function (or constant) is declared inside of that same test file, and as long as the common code is sufficiently large. For readability purposes, if the common code is less than four lines long (as an arbitrary reasonable estimate), it will be clearer if it's left repeated in each test case.
+
+Let's define an example in which a SDK client might need several lines of setup in two specific tests:
+
+```ts
+describe("some group of functionalities", function() {
+  let client;
+
+  beforeEach(function() {
+    client = new Client();
+  });
+
+  it("should test A with a proper setup", function() {
+    client.prepare({
+      propertyA: 1,
+      propertyB: 2,
+      propertyC: 3,
+      propertyD: 4,
+    });
+
+    const result = await client.A();
+    assert.ok(result.value);
+  });
+
+  it("should test B with a proper setup", function() {
+    client.prepare({
+      propertyA: 1,
+      propertyB: 2,
+      propertyC: 3,
+      propertyD: 4,
+    });
+
+    const result = await client.B();
+    assert.ok(result.value);
+  });
+});
+```
+
+It is possible to move the common setup code to a function in the same file, as in the following example code:
+
+```ts
+describe("some group of functionalities", function() {
+  let client: Client;
+
+  beforeEach(function() {
+    client = new Client();
+  });
+
+  function defaultPrepareClient(client: Client) {
+    client.prepare({
+      propertyA: 1,
+      propertyB: 2,
+      propertyC: 3,
+      propertyD: 4,
+    });
+  }
+
+  it("should test A with a proper setup", function() {
```

Comment:
Needs to be async? Same for the test below.

Reply:
I've addressed all of the async functions now! I think.